### DISCUSSION TITLE
Insert struct with slice or other struct

### DISCUSSION THREAD

**Discussion by noahfraiture** [jackc/pgx#2161] [Q&A]
Hello,
I'm trying to insert some struct, the one I have is this one :
```
type Card struct {
	ID           int
	Level        int
	Info         string
	Tasks        [2]string
	Achievements []string
	Dungeons     [3]string
	Spell        string
}
```
It contains basic type, array and slices. I tried to use `CopyFromRows` but I can't make it works.
Here's my code 
```
func tableCard() error {
	_, err := dbpool.Exec(context.Background(),
		`CREATE TABLE IF NOT EXISTS cards(
			id INTEGER PRIMARY KEY,
			level INTEGER NOT NULL,
			info TEXT,
			tasks TEXT[] CHECK (array_length(tasks, 1) <= 2),
			achievements TEXT,
			dungeons TEXT[] CHECK (array_length(dungeons, 1) <= 3),
			spell TEXT
		);`,
	)
	return err
}

func insertCards(cards []parser.Card) error {
	rows := [][]any{}
	for _, card := range cards {
		tasksArray := pgtype.Array[string]{Elements: card.Tasks[:]}
		achievementsArray := pgtype.Array[string]{Elements: card.Achievements[:]}
		dungeonsArray := pgtype.Array[string]{Elements: card.Dungeons[:]}
		rows = append(rows, []any{
			card.ID,
			card.Level,
			card.Info,
			tasksArray,
			achievementsArray,
			dungeonsArray,
			card.Spell,
		})
	}
	_, err := dbpool.CopyFrom(context.Background(), pgx.Identifier{"cards"}, []string{
		"id",
		"level",
		"info",
		"tasks",
		"achievements",
		"dungeons",
		"spell",
	}, pgx.CopyFromRows(rows))
	return err
}
```
As you can see I tried using pgtype, I also tried specifying more field of pgtype.Array, but that doesn't seem to be the solution. I could simply make a big query string by appending the value, but it's not really pretty neither safe. What are your recommendations ?

**Comment by jackc** [comment]
I don't think you should need to use pgtype.Array. You should be able to pass slices directly.

Also, you didn't say what the error message was.

**Comment by noahfraiture** [comment]
Yes sorry. It's a simple failed to encoded, because there's a mismatch between string and binary

-------- Message d'origine --------

**Comment by jackc** [comment]
That shouldn't be happening. Can you provide a runnable reproduction case?

**Reply by noahfraiture** [reply]
Sorry I can't easily isolate a test. There isn't more than what I sent. 
I solve the issue by removing the array and putting a separating char for now. You gave me the information that it should work like that, so that's already good information, thank you ! 
I will close this discussion since it should work, and it's a not a problem anymore for me
